prov/efa/src/rxr/rxr_pkt_type_rts.c

```diff
@@ -0,0 +1,783 @@
+	uint64_t tag = ~0;
+	char *rma_hdr;
+	char *data;
+	size_t data_size;
+
+	/*
+	 * rma is one sided operation, match is not expected
+	 * we need to create a rx entry upon receiving a rts
+	 */
+	rx_entry = rxr_ep_get_rx_entry(ep, NULL, 0, tag, 0, NULL, pkt_entry->addr, ofi_op_write, 0);
+	if (OFI_UNLIKELY(!rx_entry)) {
+		FI_WARN(&rxr_prov, FI_LOG_CQ,
+			"RX entries exhausted.\n");
+		efa_eq_write_error(&ep->util_ep, FI_ENOBUFS, -FI_ENOBUFS);
+		return -FI_ENOBUFS;
+	}
+
+	rx_entry->bytes_done = 0;
+
+	rts_hdr = rxr_get_rts_hdr(pkt_entry->pkt);
+	rma_hdr = rxr_pkt_read_rts_base_hdr(ep, rx_entry, pkt_entry);
+	data = rxr_pkt_read_rts_rma_hdr(ep, rx_entry, pkt_entry, rma_hdr);
+	data_size = rxr_get_rts_data_size(ep, rts_hdr);
+	return rxr_pkt_handle_rts_with_data(ep, rx_entry,
+					    pkt_entry, data,
+					    data_size);
+}
+
+static
+int rxr_pkt_proc_read_rts(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
+{
+	struct rxr_rx_entry *rx_entry;
+	struct rxr_tx_entry *tx_entry;
+	uint64_t tag = ~0;
+	int err = 0;
+	char *hdr;
+	struct rxr_pkt_rts_read_hdr *read_info;
+	/*
+	 * rma is one sided operation, match is not expected
+	 * we need to create a rx entry upon receiving a rts
+	 */
+	rx_entry = rxr_ep_get_rx_entry(ep, NULL, 0, tag, 0, NULL, pkt_entry->addr, ofi_op_read_rsp, 0);
+	if (OFI_UNLIKELY(!rx_entry)) {
+		FI_WARN(&rxr_prov, FI_LOG_CQ,
+			"RX entries exhausted.\n");
+		efa_eq_write_error(&ep->util_ep, FI_ENOBUFS, -FI_ENOBUFS);
+		return -FI_ENOBUFS;
+	}
+
+	rx_entry->bytes_done = 0;
+
+	hdr = (char *)rxr_pkt_read_rts_base_hdr(ep, rx_entry, pkt_entry);
+	hdr = (char *)rxr_pkt_read_rts_rma_hdr(ep, rx_entry, pkt_entry, hdr);
+	read_info = (struct rxr_pkt_rts_read_hdr *)hdr;
+
+	rx_entry->rma_initiator_rx_id = read_info->rma_initiator_rx_id;
+	rx_entry->window = read_info->window;
+	assert(rx_entry->window > 0);
+
+	tx_entry = rxr_rma_alloc_readrsp_tx_entry(ep, rx_entry);
+	assert(tx_entry);
+	/* the only difference between a read response packet and
+	 * a data packet is that read response packet has remote EP tx_id
+	 * which initiator EP rx_entry need to send CTS back
+	 */
+	err = rxr_ep_post_ctrl_or_queue(ep, RXR_TX_ENTRY, tx_entry, RXR_READRSP_PKT, 0);
+	if (OFI_UNLIKELY(err)) {
+		if (rxr_cq_handle_tx_error(ep, tx_entry, err))
+			assert(0 && "failed to write err cq entry");
+		rxr_release_tx_entry(ep, tx_entry);
+		rxr_release_rx_entry(ep, rx_entry);
+	} else {
+		rx_entry->state = RXR_RX_WAIT_READ_FINISH;
+	}
+
+	rxr_pkt_entry_release_rx(ep, pkt_entry);
+	return err;
+}
+
+ssize_t rxr_pkt_proc_rts(struct rxr_ep *ep,
+			 struct rxr_pkt_entry *pkt_entry)
+{
+	struct rxr_rts_hdr *rts_hdr;
+
+	rts_hdr = rxr_get_rts_hdr(pkt_entry->pkt);
+
+	if (rts_hdr->flags & RXR_READ_REQ)
+		return rxr_pkt_proc_read_rts(ep, pkt_entry);
+
+	if (rts_hdr->flags & RXR_WRITE)
+		return rxr_pkt_proc_write_rts(ep, pkt_entry);
+
+	return rxr_pkt_proc_msg_rts(ep, pkt_entry);
+}
+
+void rxr_pkt_handle_rts_recv(struct rxr_ep *ep,
```

Comment:
Not sure why this changed to `rxr_pkt_handle_rts_recv`? `rxr_pkt_handle_rts` seems clearer.

Reply:
there is rxr_pkt_handle_rts_send(), so I changed it to rxr_pkt_handle_rts_recv() to be symmetrical.